### PR TITLE
feature/tests - Example of using Jest for unit tests

### PR DESCRIPTION
**Pull request by adamburmister**
To: @zackargyle 

Just passing along at this stage for your thoughts - quite a bit refactor to named exports (exporting functions/consts directly rather than the antipattern of object exports)

It turns out Jest isn't really that great... I'm having a hell of a time mocking out click handlers and such on my simple test case, so may explore other options shortly.

### CODE REVIEW

File: src/util/PinConfig.js [hunk 1]
```diff
@@ -1,17 +1,26 @@
...
-    [Const.COUNT.BUTTON]: 0,
-    [Const.COUNT.FOLLOW]: 0,
-    [Const.COUNT.PIN_SMALL]: 0,
-    [Const.COUNT.PIN_MEDIUM]: 0,
-    [Const.COUNT.PIN_LARGE]: 0,
-    [Const.COUNT.PROFILE]: 0,
-    [Const.COUNT.BOARD]: 0
-};
\ No newline at end of file
+export const guid = generateGuid();
```
Comment (adamburmister) [inline src/util/PinConfig.js]: Not quite sure why we've got this GUID? I think this is generated once per run, so may move into Constants

File: package.json [hunk 2]
```diff
@@ -31,7 +35,30 @@
     "build:lib": "babel src --out-dir lib",
     "build:dev": "webpack lib/main.js dist/react-pinterest.js --config webpack.config.base.js",
     "build:prod": "webpack lib/main.js dist/react-pinterest.min.js --config webpack.config.prod.js",
-    "prepublish": "npm run clean && npm run build"
+    "prepublish": "npm run clean && npm run build",
+    "test": "jest",
+    "test:watch": "npm run test -- --watch"
```
Comment (zackargyle) [inline package.json]: Was there supposed to be another field?

File: src/components/PinterestBase.js [hunk 3]
```diff
@@ -1,38 +1,38 @@
...
- * 
+ *
  */
 export default class PinterestBase extends React.Component {
 
     constructor(props) {
         super(props)
         this.debounceInitialLogging();
-        i18n.setLang(this.props.lang || 'en');
+        i18n.lang = this.props.lang;
```
Comment (zackargyle) [inline src/components/PinterestBase.js:17]: I need to redo all the i18n stuff. The translation portion is fine, but determining what language to use pretty sure is not working correctly.

File: src/util/PinConfig.js (hunk 1 above)
Comment (zackargyle) [inline src/util/PinConfig.js]: The guid is used for all of our logging calls. If they travel to Pinterest from one of the widgets, the guid is placed as a query parameter and then we can link it back to the user/domain/widget they came from. Could definitely go in Constants. We could also move counts onto PinterestBase as a static value like you did with `initialized` and get rid of this whole file. :+1:  Not this PR, it's already too big. lol

File: src/util/PinConst.js [hunk 4]
```diff
@@ -1,29 +1,220 @@
...
+    BUTTON: 'button_count',
+    FOLLOW: 'follow_count',
+    PROFILE: 'profile_count',
+    BOARD: 'board_count',
+    PIN_SMALL: 'pin_count',
+    PIN_MEDIUM: 'pin_count_medium',
+    PIN_LARGE: 'pin_count_large'
 };
+
+export const I18N_DICTIONARY = Object.freeze({
```
Comment (zackargyle) [inline src/util/PinConst.js]: Better place for this. :+1:

File: src/util/i18n.js [hunk 5]
```diff
@@ -1,18 +1,52 @@
...
+        this._dictionary = dictionary;
+    }
+
+    /**
+     * Create a new i18n class to translate strings
+     * @param {Object} dictionary
+     * @param {string} language Initial language for the dictionary
+     */
+    constructor(dictionary, language) {
+        this._lang = language || DEFAULT_LANGUAGE;
```
Comment (zackargyle) [inline src/util/i18n.js:46]: Don't need the `|| DEFAULT_LANGUAGE` since that's exactly what you're passing in, and it's a singleton.

**Comment by zackargyle** [comment]
In general, I think the PR looks good. Does seem like a lot of shimming is happening in the test cases, but I haven't looked into testing enough to know better ways off the top of my head to handle event-mocks. We can explore more after break.

**Comment by zackargyle** [comment]
Also, can you squash the commits? 13 is a little excessive ;)

File: src/components/PinterestBase.js (hunk 3 above)
Comment (adamburmister) [inline src/components/PinterestBase.js:17]: Ok I won't worry too much about it then

File: src/util/i18n.js (hunk 5 above)
Comment (adamburmister) [inline src/util/i18n.js:46]: I meant to export the class too which made it possible to test the default language via constructor injection

**Comment by adamburmister** [comment]
I'm not usually a commit squasher - what's the syntax? `git rebase -i HEAD~13` on the current branch?

**Comment by zackargyle** [comment]
Yeah. When it pops up in the editor change all of the commits from `pick` to `squash`, except for the one you want to act as the commit bundle.

**Comment by adamburmister** [comment]
@zackargyle probably ready for another review

File: src/components/__tests__/PinterestAnchor.test.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
+jest.dontMock('../PinterestAnchor');
+
+import React from 'react';
+import ReactDOM from 'react-dom';
+import TestUtils from 'react-addons-test-utils';
+
+const shallowRenderer = TestUtils.createRenderer();
+
+// Require the component to test
+const PinterestAnchor = require('../PinterestAnchor').default;
```
Comment (zackargyle) [inline src/components/__tests__/PinterestAnchor.test.js:10]: Can we move away from `require` now?

Comment (adamburmister) [inline src/components/__tests__/PinterestAnchor.test.js:10]: Good question... need to retry that.

File: src/components/PinterestBase.js (hunk 3 above)
Comment (adamburmister) [inline src/components/PinterestBase.js:17]: @zackargyle The answer is "no" ... "because of Jest"
https://facebook.github.io/jest/docs/tutorial-react.html

**Comment by zackargyle** [comment]
Looks good to me!

**Pull request merged**